pulley/src/regs.rs

```diff
@@ -0,0 +1,180 @@
+            }
+
+            #[cfg(feature = "arbitrary")]
+            impl<'a> arbitrary::Arbitrary<'a> for $name {
+                fn arbitrary(u: &mut arbitrary::Unstructured<'a>) -> arbitrary::Result<Self> {
+                    let index = u.int_in_range(Self::RANGE.start..=Self::RANGE.end - 1)?;
+                    Ok(Self(index))
+                }
+            }
+        )*
+    }
+}
+
+define_registers! {
+    /// An `x` register: integers.
+    pub struct XReg = 0..37;
+
+    /// An `f` register: floats.
+    pub struct FReg = 0..32;
+
+    /// A `v` register: vectors.
+    pub struct VReg = 0..32;
+}
+
+/// Any register, regardless of class.
+///
+/// Never appears inside an instruction -- instructions always name a particular
+/// class of register -- but this is useful for testing and things like that.
+#[allow(missing_docs)]
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum AnyReg {
+    X(XReg),
+    F(FReg),
+    V(VReg),
+}
+
+impl From<XReg> for AnyReg {
+    fn from(x: XReg) -> Self {
+        Self::X(x)
+    }
+}
+
+impl From<FReg> for AnyReg {
+    fn from(f: FReg) -> Self {
+        Self::F(f)
+    }
+}
+
+impl From<VReg> for AnyReg {
+    fn from(v: VReg) -> Self {
+        Self::V(v)
+    }
+}
+
+impl fmt::Display for AnyReg {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            AnyReg::X(r) => fmt::Display::fmt(r, f),
+            AnyReg::F(r) => fmt::Display::fmt(r, f),
+            AnyReg::V(r) => fmt::Display::fmt(r, f),
+        }
+    }
+}
+
+impl fmt::Debug for AnyReg {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> core::fmt::Result {
+        fmt::Display::fmt(self, f)
+    }
+}
+
+#[cfg(feature = "arbitrary")]
+impl<'a> arbitrary::Arbitrary<'a> for AnyReg {
+    fn arbitrary(u: &mut arbitrary::Unstructured<'a>) -> arbitrary::Result<Self> {
+        match u.int_in_range(0..=2)? {
+            0 => Ok(AnyReg::X(u.arbitrary()?)),
+            1 => Ok(AnyReg::F(u.arbitrary()?)),
+            2 => Ok(AnyReg::V(u.arbitrary()?)),
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl XReg {
+    /// The valid special register range.
+    pub const SPECIAL_RANGE: Range<u8> = 32..37;
+
+    /// The special `sp` stack pointer register.
+    pub const SP: Self = Self(32);
+
+    /// The special `lr` link register.
+    pub const LR: Self = Self(33);
+
+    /// The special `fp` frame pointer register.
+    pub const FP: Self = Self(34);
+
+    /// The special `spilltmp0` scratch register.
+    pub const SPILL_TMP_0: Self = Self(35);
+
+    /// The special `spilltmp1` scratch register.
+    pub const SPILL_TMP_1: Self = Self(36);
+
+    /// Is this `x` register a special register?
+    pub fn is_special(&self) -> bool {
+        self.0 >= Self::SPECIAL_RANGE.start
+    }
+}
+
+impl fmt::Display for XReg {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            x if *x == Self::SP => write!(f, "sp"),
```

Comment:
Is it actually possible to do `Self::SP => write!(f, "sp")` here? I forget what we can do with constants-in-patterns...

Reply:
Tried that first, unfortunately it isn't allowed :(